### PR TITLE
Add a check if randspg binary is installed

### LINKED ISSUE #670: the randspg job has no warning that it fails for windows
In the process of getting contrib switched over to using pyiron/actions, I discovered that the [RandSpg job](https://github.com/pyiron/pyiron_contrib/blob/b6a1fb33385fcb1c87e41b69477b7038ef07ff8c/pyiron_contrib/atomistics/randspg/randspg.py#L26) depends on `randspg 0.0.1` --- a package unavailable for windows.

Currently this dependency is in `.ci_support/environment.yml` but _not_ in `setup.py` or the [conda-forge feedstock dependencies](https://github.com/conda-forge/pyiron_contrib-feedstock/blob/main/recipe/meta.yaml).

In #669 I expect this to work fine since this class doesn't actually have any unit tests, so having it in the notebook env should be enough to keep the notebook tests running (which are on ubuntu-latest, so we have access to the module).

I am concerned that this job will fail in a not-clean way for Windows users.

@jan-janssen, @pmrv, both of you have made commits to this job class; does one of you want to wrap it in an import alarm or some other solution?

### PR DESCRIPTION
**Pull request by pmrv**
Fixes #670

### CODE REVIEW

**Review by liamhuber — COMMENTED** [review]
typo, but lgtm

File: pyiron_contrib/atomistics/randspg/randspg.py [hunk 1]
```diff
@@ -69,6 +70,10 @@ def _get_structure(self, frame, wrap_atoms=True):
         return self._structure_storage._get_structure(frame=frame, wrap_atoms=wrap_atoms)
 
 
+    def validate_ready_to_run(self):
+        is shutil.which("randSpg") is not None:
```
[the remark this answers is not in the document]
Comment (pmrv) [reply pyiron_contrib/atomistics/randspg/randspg.py]: Derp.

**Comment by pmrv** [comment]
I guess since, we can't have it in setup.py this might make still sense with the updated conda package.

**Comment by pmrv** [comment]
I guess since, we can't have it in setup.py this might make still sense with the updated conda package.

**Comment by liamhuber** [comment]
> I guess since, we can't have it in setup.py this might make still sense with the updated conda package.

👍👍 agreed

**Comment by jan-janssen** [comment]
> I guess since, we can't have it in setup.py this might make still sense with the updated conda package.

Does `shutil.which()` find the windows executable?

**Comment by pmrv** [comment]
[Should] be, as long as it's installed in the `PATH`.

File: pyiron_contrib/atomistics/randspg/randspg.py [hunk 2]
```diff
@@ -69,6 +70,10 @@ def _get_structure(self, frame, wrap_atoms=True):
         return self._structure_storage._get_structure(frame=frame, wrap_atoms=wrap_atoms)
 
 
+    def validate_ready_to_run(self):
+        if shutil.which("randSpg") is not None:
```
Comment (pmrv) [inline pyiron_contrib/atomistics/randspg/randspg.py:74]:
```suggestion
        if shutil.which("randSpg") is None:
```

**Comment by jan-janssen** [comment]
> [Should](https://docs.python.org/3/library/shutil.html?highlight=shutil#shutil.which) be, as long as it's installed in the `PATH`.

Work fine - at least in the test repository https://github.com/pyiron-dev/randspg-pyiron-win/pull/1 . I guess the only issue is if somebody compiled randspg themselves and just put the `run_randspg_0.0.1.sh` script in their resource directory it is possible that this would still raise an error. This case is less likely for randspg but more for sphinx https://github.com/pyiron/pyiron_atomistics/pull/1030

**Comment by jan-janssen** [comment]
Can I deactivate this check from the users side? While it is less frequent for randspg I just like to have a general solution what we do when the executable is not available. In particular as I see no time pressure for merging this pull request as `randspg` is now available for windows as well.

**Comment by jan-janssen** [comment]
> I'd rather not allow the perfect to be the enemy of the good and leave this PR to languish indefinitely. Could we just update the error message with something to the effect of "If you already installed randspg yourself somewhere, please just make sure `shutil.which` can find it, e.g. by adding its location to your `PATH` on a Windows machine"?

The issue to me that this is a hard fail. So in case the user has to load a module to have access to `randspg` then they have to load the module by default to prevent pyiron from failing hard. I am also fine if there is a flag on the job level to just disable the check manually and have it being raised by default.

**Comment by liamhuber** [comment]
> The issue to me that this is a hard fail. So in case the user has to load a module to have access to `randspg` then they have to load the module by default to prevent pyiron from failing hard.

Isn't this exactly desirable behaviour? If the user doesn't have access to `randspg`, and they try to run a `RandSpg` job, don't we precisely want a hard and clean failure?

The time we wouldn't want such a failure is if the user has access to a `randspg` executable but it _isn't_ findable by `shutil.which`. Does such a case exist? E.g. in the case of windows, `shutil.which` will find anything that is in `PATH` or in the local directory (per the docs Marvin posted), but in this case the user also has no access to `randspg`!

Can you clarify the edge case you're looking to avoid here?

**Comment by jan-janssen** [comment]
I want a solution which can also be applied to https://github.com/pyiron/pyiron_atomistics/pull/1030 . But in the case of sphinx we rarely install it in the `PATH` but more likely we use `module load` to extend the path only at run time.

**Comment by liamhuber** [comment]
> I want a solution which can also be applied to [pyiron/pyiron_atomistics#1030](https://github.com/pyiron/pyiron_atomistics/pull/1030) . But in the case of sphinx we rarely install it in the `PATH` but more likely we use `module load` to extend the path only at run time.

...You want a solution here that will solve another problem in another repository. Jan, it's a damned error message. This is madness.

I'll concede that it would be great to extend the message letting users know how to fix it in the one-in-a-million chance that they get into the edge case of _having_ a `randSpg` accessible but not findable by `shutil.which("randSpg")` -- e.g. because they have it some non-canonical name. But this is otherwise a perfectly good fix to the problem we're presented with here. If we later get a solution that solves [pyiron/pyiron_atomistics#1030](https://github.com/pyiron/pyiron_atomistics/pull/1030) and we also want to apply it here to improve consistency, then I'm all for that. But I see need to wait for that in order to fix this here and now.

**Comment by jan-janssen** [comment]
> ...You want a solution here that will solve another problem in another repository. Jan, it's a damned error message. This is madness.

As discussed in https://github.com/pyiron/team/discussions/169#discussioncomment-5639949 and in https://github.com/pyiron/team/discussions/172 my current main concern is that pyiron is not maintainable. My aim is to write code which is minimal number of lines of code is addressing the majority of a given challenge. The way I see the current discussion is that the have two pull requests which try to implement the same feature for two different codes: 
* https://github.com/pyiron/pyiron_contrib/pull/671
* https://github.com/pyiron/pyiron_atomistics/pull/1030

None of these changes is urgent. So I want us to take the time to come up with a solution which can be used for both codes. Just like we have the import alarm for python codes https://github.com/pyiron/pyiron_base/blob/main/pyiron_base/utils/error.py we need something similar for executables. But I do not see the benefit of merging this pull request, if we agree that this solution only applies to the subset of codes which can be installed via conda and does not work for codes which are loaded as separate modules, like sphinx, VASP and so on. 

Even worse, I have the feeling that we spend more time discussing the five lines in this pull request, rather than progressing towards a general solution.

**Comment by liamhuber** [comment]
> But I do not see the benefit of merging this pull request

There are only two possibilities:

1) We merge this
- pyiron_contrib behaves slightly cleaner without any impact on the rest of the codebase 
- Once a universal solution is found for [pyiron_atomistics #1030](https://github.com/pyiron/pyiron_atomistics/pull/1030), we come implement it here

2) We close this
- pyiron_contrib stays with a bad failure for anyone who uses this job after a PyPi installation
- Once a universal solution is found for [pyiron_atomistics #1030](https://github.com/pyiron/pyiron_atomistics/pull/1030), we come implement it here

I honestly do not understand how (2) is preferable to (1).

> Even worse, I have the feeling that we spend more time discussing the five lines in this pull request, rather than progressing towards a general solution.

💯

**Comment by jan-janssen** [comment]
> I honestly do not understand how (2) is preferable to (1).

As it is `pyiron_contrib` I am fine with (1), for the pull request in `pyiron_atomistics` I am not fine with merging the same solution.

**Comment by liamhuber** [comment]
> in `pyiron_atomistics` I am not fine with merging the same solution.

Here I agree 100%, although my rationale there is not anything with code length, but rather your point about those jobs calling shell scripts which should be free to modify the path only _after_ this validation step is done -- it's simply not the right solution for that more complex problem.

**Comment by pmrv** [comment]
The immediate grounds is to raise a logger warning, if the executable is not found---for the case that the run script of the job makes the executable available, but it's not available in `PATH`.  That case is generally intractable, though.  A universal solution cannot exist.  `pyiron` can never now what the run script does to run or not.  I could write a script that downloads and compiles a missing code, even.  So the only sane thing imo is not to try.  But we can also not fail hard in the case @jan-janssen describes, because it will break in some circumstances without users being able remedy it.  For contrib I have no strong feelings, but for https://github.com/pyiron/pyiron_atomistics/pull/1030 I say let's just change the error to a warning.  Users then have two locations that clearly state a binary is missing: the warning and `job['error.msg']`, that ought to be enough.

**Pull request merged**